### ISSUE TITLE
[SOLVED]  chectl server:  commands return "Error Unauthorized" after update

### ISSUE THREAD

**Issue by Mbd06b** [eclipse-che/che#19642]
I've been using chectl successfully for a while. 
chectl updated itself  from 7.28.0 to 7.29.0 when I was going to perform a chectl server:update. 

chectl version
chectl/7.29.0 linux-x64 node-v12.22.1

almost every chectl che commands that I use that interact with che returns:
"Error: Unauthorized"

without any other details.

any ideas? 

I use microk8s on ubuntu 18 LTS.
 kubectl version
Client Version: version.Info{Major:"1", Minor:"21+", GitVersion:"v1.21.0-3+121713cef81e03", GitCommit:"121713cef81e0316fb97af2111e1d3aeac3fd14a", GitTreeState:"clean", BuildDate:"2021-04-11T13:30:54Z", GoVersion:"go1.16.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"21+", GitVersion:"v1.21.0-3+121713cef81e03", GitCommit:"121713cef81e0316fb97af2111e1d3aeac3fd14a", GitTreeState:"clean", BuildDate:"2021-04-11T13:26:38Z", GoVersion:"go1.16.3", Compiler:"gc", Platform:"linux/amd64"}

**Comment by l0rd** [comment]
@Mbd06b can you please provide the full command, output and log of the command that is failing? The previous and working version of chectl was 7.28.0?

**Comment by tolusha** [comment]
@Mbd06b 
Do you have any updates on this issue?

**Comment by Mbd06b** [comment]
![Capture](https://user-images.githubusercontent.com/12649222/116264145-ab152500-a73f-11eb-8e6e-54a2b3ef6870.PNG)

**Comment by tolusha** [comment]
Could you check `~/.cache/chectl/error.log` if there is anything interesting there?

**Comment by tolusha** [comment]
It might be related to https://github.com/eclipse/che/issues/15331

**Comment by Mbd06b** [comment]
Here's a sample from ~/.cache/chetcl/error.log   .the messages are all following similar pattern... 

![error_log](https://user-images.githubusercontent.com/12649222/116441679-44633a80-a817-11eb-9b3f-e76f20bcd171.png)

chenamespace doesn't seem to do anything...

![chenamespace](https://user-images.githubusercontent.com/12649222/116441885-7d9baa80-a817-11eb-8cc8-5779a437905e.png)

**Comment by tolusha** [comment]
I dig in the code and I don't see anything strange. 
I simply tries to list `checlusters` objects in namespace.

Could you try the previous version if it still works (ignore chectl update)?

```bash
cd ~/.local/share/chectl/client/7.28.0/bin
./run server:status
```

**Comment by Mbd06b** [comment]
> ```shell
> erver:status
> ```

Are there instructions on how to downgrade chectl?

**Comment by Mbd06b** [comment]
My best guess at this point is that there is some kind of disagreement between kubectl and chectl. 

I assume chectl uses chectl to do it's work, but every time it attempts to do something kubernetes is rejecting chectl's requests.  

'KubeHelper.wrapK8sClientError()'

**Comment by tolusha** [comment]
To remove chectl completely:
1. `whereis chectl`
2. `sudo rm -rf /usr/local/bin/chectl /usr/local/lib/chectl` (I guess the same on your side)

To install from the scratch:
3.  `bash <(curl -sL  https://www.eclipse.org/che/chectl/)`

**Comment by Mbd06b** [comment]
This is the key: Why is this failing? 
`2021-05-16T19:33:32.281Z Error: Unauthorized
2021-05-16T19:33:32.281Z     at KubeHelper.wrapK8sClientError (/usr/local/lib/chectl/lib/api/kube.js:2784:20)
2021-05-16T19:33:32.281Z     at KubeHelper.<anonymous> (/usr/local/lib/chectl/lib/api/kube.js:1742:28)
2021-05-16T19:33:32.281Z     at Generator.throw (<anonymous>)
2021-05-16T19:33:32.281Z     at rejected (/usr/local/lib/chectl/node_modules/tslib/tslib.js:115:69)
`

**Comment by tolusha** [comment]
@Mbd06b 
It is hard to say.
It would be nice if you debug in your environment.

```bash
git clone https://github.com/che-incubator/chectl.git
yarn
cd bin
./debug
```

Use VS Code and `Attach to remote` debug configuration or any other IDE.

**Comment by Mbd06b** [comment]
@tolusha  I've been working on setting up a remote debug.  I've not done a remote debug before so it will take me a little time to set it up and try it out.

**Comment by Mbd06b** [comment]
I've got a port conflict with running chectl debugger on the remote server I'm trying to fix. 
![Capture](https://user-images.githubusercontent.com/12649222/119271303-a456c200-bbc6-11eb-9501-998999ddf622.PNG)

**Comment by Mbd06b** [comment]
ok, I'm able to run the ./debug on the remote server, just  needed to specify to node inspect port in  in chectl/bin/debug 
![image](https://user-images.githubusercontent.com/12649222/119272249-b33f7380-bbca-11eb-9130-0e76bd54d0a1.png)

**Comment by Mbd06b** [comment]
So far this is the error that I can see from stepping through debug. 
`Error: Unauthorized\n    at KubeHelper.wrapK8sClientError (/home/user/chectl/src/api/kube.ts:2617:74)\n    at KubeHelper.<anonymous> (/home/user/chectl/src/api/kube.ts:1646:18)\n    at Generator.throw (<anonymous>)\n    at rejected (/home/user/chectl/src/api/kube.ts:15:65)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)`

**Comment by Mbd06b** [comment]
So I'm a bit stuck at the moment as to how to set a breakpoint (at that location) on the remote node application itself.  
during the debug the source of the files in VS code look like this...  (\<node internals>\internal\per_context\primoridals.js)

**Comment by tolusha** [comment]
Everything looks find. I see a breakpoint mark on 1626 line. 
So, debugger must stop at that line when you run it.

**Comment by Mbd06b** [comment]
Ok, I figured it out, and it's not chectl's fault (however chectl errors had no clues to the feedback)

 What was happening was that my kube config credentials were not being read properly, this was preventing not only chectl from working, but helm was throwing errors too. 


This config is found in ~/.kube/config 
These credentials need to be set with the right file access permissions, and with a cluster context that will allow 3rd party tools like chectl and helm to authenticate against the cluster to perform operations. 

I haven't been able to figure out what exactly triggered my problem (maybe a microk8s update), 
and I'm not exactly sure how I fixed it.  I just remember taking the Kube Config from my running kubernetes cluster and dumping it out with a "raw >"  command that looked like that, and then deleting a duplicated configuration which got everything working again for chectl and helm. 

hopefully that's enough clues for anyone to Google for them to figure out what went wrong with them.
